### PR TITLE
Adding precision issue note docs for `functional.interpolate`

### LINKED ISSUE #104157: `interpolate` bicubic mode returns unexpected value on CUDA with float16
### 🐛 Describe the bug

Hello, I've encountered an issue while training a model that produces unusual output. Upon investigation, I discovered that the anomaly originates from an unexpected result from the 'interpolate' function.

```python
import torch
from torch.nn.functional import interpolate

def get_grads(dtype, size):
    input = torch.randn((2, 2, 1, 1), dtype=dtype, requires_grad=True).cuda()

    output = interpolate(input, size=size, mode='bicubic', align_corners=True)

    grad_outputs = [torch.ones_like(output)]
    grads = torch.autograd.grad([output], [input], grad_outputs)
    return grads[0]

print(get_grads(torch.float32, (128, 128)).flatten())
print(get_grads(torch.float16, (128, 128)).flatten())

print(get_grads(torch.float32, (64, 64)).flatten())
print(get_grads(torch.float16, (64, 64)).flatten())

print(get_grads(torch.float32, (32, 32)).flatten())
print(get_grads(torch.float16, (32, 32)).flatten())
```

The output is as follows:

```bash
tensor([16384., 16384., 16384., 16384.], device='cuda:0')
tensor([2048., 2048., 2048., 2048.], device='cuda:0', dtype=torch.float16)
tensor([4096., 4096., 4096., 4096.], device='cuda:0')
tensor([2048., 2048., 2048., 2048.], device='cuda:0', dtype=torch.float16)
tensor([1024., 1024., 1024., 1024.], device='cuda:0')
tensor([1024., 1024., 1024., 1024.], device='cuda:0', dtype=torch.float16)
```

Could this be a bug? It appears that the maximum value of the gradient is capped at 2048 for the float16 data type. If this is not a bug, could you please tell me why this is happening?

Furthermore, I'm wondering if you could implement a warning to alert users when this issue occurs.

### Versions

Collecting environment information...
PyTorch version: 2.0.0a0+gitc263bd4
Is debug build: False
CUDA used to build PyTorch: 11.2
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 7.5.0
Clang version: Could not collect
CMake version: version 3.26.3
Libc version: glibc-2.17

Python version: 3.8.5 (default, Sep  4 2020, 07:30:14)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-3.10.0-693.el7.x86_64-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.2.67
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB
GPU 4: Tesla V100-SXM2-32GB
GPU 5: Tesla V100-SXM2-32GB
GPU 6: Tesla V100-SXM2-32GB
GPU 7: Tesla V100-SXM2-32GB

Nvidia driver version: 460.73.01
cuDNN version: /usr/local/cuda-9.0/lib64/libcudnn.so.7.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                48
On-line CPU(s) list:   0-47
Thread(s) per core:    2
Core(s) per socket:    12
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Gold 6146 CPU @ 3.20GHz
Stepping:              4
CPU MHz:               3200.000
BogoMIPS:              6400.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              25344K
NUMA node0 CPU(s):     0-11,24-35
NUMA node1 CPU(s):     12-23,36-47
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cdp_l3 intel_pt tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm arat pln pts

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.0.0a0+gitc263bd4
[pip3] torchvision==0.15.1a0
[conda] magma-cuda112             2.5.2                         1    https://pkg.sensetime.com/repository/conda-tsinghua-cloud/pytorch
[conda] mkl                       2023.1.0                 pypi_0    pypi
[conda] mkl-include               2023.1.0                 pypi_0    pypi
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] torch                     2.0.0a0+gitc263bd4          pypi_0    pypi
[conda] torchvision               0.15.1a0                 pypi_0    pypi

### PR DESCRIPTION
**Pull request by yewentao256**
Fixes #104157

### CODE REVIEW

**Comment by lezcano** [comment]
There's a number of unrelated changes in the PR, perhaps an incorrect rebase?

**Comment by yewentao256** [comment]
> There's a number of unrelated changes in the PR, perhaps an incorrect rebase?

It seems so... Should I close this pr and fixes that issue, and make a pr again?

**Comment by vfdev-5** [comment]
> > There's a number of unrelated changes in the PR, perhaps an incorrect rebase?
> 
> It seems so... Should I close this pr and fixes that issue, and make a pr again?

no, you can just rebase correctly and push force to your fork

File: torch/nn/functional.py [hunk 1]
```diff
@@ -3878,6 +3878,12 @@ def interpolate(input: Tensor, size: Optional[int] = None, scale_factor: Optiona
         backward compatibility.
         Mode ``mode='nearest'`` matches buggy OpenCV's ``INTER_NEAREST`` interpolation algorithm.
 
+    .. note::
+        The use of ``float16`` data type in the backward pass of the upsample operation, specifically
+        for modes such as ``['linear', 'bilinear', 'bicubic', 'trilinear', 'area']``, could potentially
+        lead to precision issues. For instance, any result that exceeds 2048 will be truncated.
+        For more details, please refer to the discussion in `issue#104157 <https://github.com/pytorch/pytorch/issues/104157>`_.
```
Comment (vfdev-5) [inline torch/nn/functional.py:3885]: I do not think we should mention 2048 here as it is specific to your repro example from the issue.

**Comment by yewentao256** [comment]
> > > There's a number of unrelated changes in the PR, perhaps an incorrect rebase?
> > 
> > 
> > It seems so... Should I close this pr and fixes that issue, and make a pr again?
> 
> no, you can just rebase correctly and push force to your fork

It seems I've made another incorrect rebase 😂. I just rebased to the main branch of the PyTorch repository and pushed to my fork. I sincerely apologize for the mistake! Here's the command sequence I used:
```bash
git remote -v
myfork  git@github.com:yewentao256/pytorch.git (fetch)
myfork  git@github.com:yewentao256/pytorch.git (push)
origin  git@github.com:pytorch/pytorch.git (fetch)
origin  git@github.com:pytorch/pytorch.git (push)
(base) PS C:\Users\Peter\Desktop\pytorch> git rebase origin main
dropping 16d268e280514b61f8b731893b7bc6270c2e3b7b Fix comment error in TensorIterator.cpp (#100227) -- patch contents already upstream
Successfully rebased and updated refs/heads/main.
(base) PS C:\Users\Peter\Desktop\pytorch> git push myfork main --force
```
Is there anything I can do to fix this problem now?

**Comment by vfdev-5** [comment]
@yewentao256 try to do the following:
```
git reset --hard HEAD~2500
git pull origin main -r
git cherry-pick e015d03
# at this point you should have a linear history + one your commit with docs update
git log --oneline --decorate --graph

git push myfork main -f
```

A better practice is to create a branch and commit to a branch instead of `main`.

**Comment by yewentao256** [comment]
> @yewentao256 try to do the following:
> 
> ```
> git reset --hard HEAD~2500
> git pull origin main -r
> git cherry-pick e015d03
> # at this point you should have a linear history + one your commit with docs update
> git log --oneline --decorate --graph
> 
> git push myfork main -f
> ```
> 
> A better practice is to create a branch and commit to a branch instead of `main`.

@vfdev-5  Thanks a lot! Solved the issue👍

**Comment by vfdev-5** [comment]
Now, please address this comment: https://github.com/pytorch/pytorch/pull/104622#discussion_r1252991212

File: torch/nn/functional.py (hunk 1 above)
Comment (yewentao256) [reply torch/nn/functional.py]: Done

**Comment by yewentao256** [comment]
> Now, please address this comment: [#104622 (comment)](https://github.com/pytorch/pytorch/pull/104622#discussion_r1252991212)

I removed the 2048 example there, is there anything else that I need to update? Please just let me know

**Comment by vfdev-5** [comment]
I would rephrase the message saying something like "gradients may be incorrect for dtype float16", I let your reword it in a better way. Also, I think this happens on CUDA only and not on CPU (can you please check?)

**Comment by yewentao256** [comment]
> I would rephrase the message saying something like "... gradients may be incorrect for dtype float16 ...", I let your reword it in a better way. IMO, the main message is "gradients can be incorrect". Also, I think this happens on CUDA only and not on CPU (can you please check?)

Yes, you are right, this issue will not happen on CPU (not supported for float16 currently).

Here's the revised content:

        The gradients may be incorrect for dtype ``float16`` on CUDA in the backward pass of the upsample when using modes ``['linear', 'bilinear', 'bicubic', 'trilinear', 'area']``. For more details, please refer to the discussion in `issue#104157 <https://github.com/pytorch/pytorch/issues/104157>`_.

If this makes sense, I will push a new commit then.

**Comment by yewentao256** [comment]
> Yes, looks better for me. I would just remove "during the backward pass" as we already talking about gradients

Done

**Review by ezyang — APPROVED** [review]
